### ISSUE TITLE
[Help Wanted] [PLT-3447] Add support for context-specific translations

### ISSUE THREAD

**Issue by esethna** [mattermost/platform#5267]
If you're interested please comment here and come [join our "Contributors" community channel](https://pre-release.mattermost.com/core/channels/tickets) on our daily build server, where you can discuss questions with community members and the Mattermost core team. For technical advice or questions, please  [join our "Developers" community channel](https://pre-release.mattermost.com/core/channels/developers).

New contributors please see our [Developer's Guide](https://docs.mattermost.com/guides/developer.html), specifically for [machine setup](https://docs.mattermost.com/developer/developer-setup.html) and for [developer workflow](https://docs.mattermost.com/developer/developer-flow.html).

----

**Notes**: [Jira ticket](https://mattermost.atlassian.net/browse/PLT-3447)

**Comment by hanzei** [comment]
@jasonblais Do you know if this problem still exists?

**Comment by jasonblais** [comment]
As far as I know, it is

**Comment by hanzei** [comment]
V2 of https://github.com/nicksnyder/go-i18n supports this. But it would be a huge effort to migrate to this version.

**Comment by jasonblais** [comment]
Updated ticket with new examples and options. Anyone in the community interested helping with this improvement?

**Comment by andylibrian** [comment]
@jasonblais @hanzei 

I'd like to see how big the effort to do this. Do you mind pointing me to the example / feature in https://github.com/nicksnyder/go-i18n V2 where we can use for this?

I looked at the documentation and didn't have idea on how to switch between `Neue` and `Neue(n)`. 
The switch seems to be based on PluralCount only, as I understand it, `"Neuen Kanal" and "Neue Gruppe"` is not based on plural count.

**Comment by jasonblais** [comment]
Thanks Andy! @hanzei wondering if you would know to point to a specific example on the library you referenced?

**Comment by hanzei** [comment]
Hi @andylibrian, thanks for you interest in this ticket. Example 1 can easily be fixed by using a plural form. 

It might be reasonable to also use plural forms for Example 2 and provide extra information via `Description`. Or the message could just be duplicated and used two different localisation in e.g. German.

**Comment by andylibrian** [comment]
@hanzei I looked at this again and it seems that plural form is supported in existing translation:

https://github.com/mattermost/mattermost-server/blob/605040c59715be1b06671e9b62de8b5a10989d70/i18n/de.json#L1033-L1039

https://github.com/mattermost/mattermost-server/blob/605040c59715be1b06671e9b62de8b5a10989d70/i18n/de.json#L1483-L1489

**Comment by hanzei** [comment]
The main advantage of v2 is that it supports explicit plural forms and a lot more. See https://godoc.org/github.com/nicksnyder/go-i18n/v2/i18n#Message Also, we can provide a `Description` that would help translators.

**Comment by andylibrian** [comment]
Right, v2 has more features. Currently some of the code depends on `goi18n.TranslateFunc` directly.

https://github.com/mattermost/mattermost-server/blob/c21ea5bc065087d12ed8407fb480b723f836339c/app/app.go#L28

And most of the translation are done by calling `utils.T`

https://github.com/mattermost/mattermost-server/blob/c21ea5bc065087d12ed8407fb480b723f836339c/app/channel.go#L21-L25

Let me know if I'm missing something.

So, I think a gradual approach to upgrade this is to create an abstraction, define our own `utils.T` which will be compatible with existing code. Then the function will call v2 localizer:

```go
localizer.MustLocalize(&i18n.LocalizeConfig{
    DefaultMessage: &i18n.Message{
        ID: "..."
        One: "...",
        Other: "...",
    },
    TemplateData: map[string]interface{}{
        "Name": "...",
        "Count": 2,
    },
    PluralCount: 2,
})
```

**Comment by hanzei** [comment]
Once challenge is  that `model.CommandArgs` uses `goi18n.TranslateFunc` directly. Given that this change be broken because of semver, I would opt for leaving all `app/command*` files and `model.CommandArgs` as it is.

Did you look into the generated translation files yet? There might be issues using the new file structure with the existing translation server.

**Comment by andylibrian** [comment]
> Did you look into the generated translation files yet? There might be issues using the new file structure with the existing translation server.

Yes, I mentioned a few of translation files in my previous comment. So, for this case, we could create our own function to load translation messages from existing json files.

Instead of:
```go
bundle.RegisterUnmarshalFunc("toml", toml.Unmarshal)
bundle.LoadMessageFile("es.toml")
```

We could probably use something like this:
```go
bundle.RegisterUnmarshalFunc("json", utils.i18nJsonUnmarshall)
bundle.LoadMessageFile("es.json")
```

And regarding semver, when you say

> I would opt for leaving all app/command* files and model.CommandArgs as it is.

Does that mean you opt for 2 ways of translating messages (v1 and v2)? And the plan would be to move v1 to v2 when breaking change is allowed (mattermost-server v6)?

**Comment by KuSh** [comment]
It seems this isn't a need or a priority anymore as my PRs are both staled :/

**Comment by M-ZubairAhmed** [comment]
@esethna  is this still relevant?

**Comment by KuSh** [comment]
Hi this is not "Up For Grabs" as I'm still maintaining my PR from last year and it progress (alas slowly)

**Comment by jasonblais** [comment]
@KuSh removed the up for grabs label for you.

I noticed you mentioned

> It seems this isn't a need or a priority anymore as my PRs are both staled :/

Do you have clarity on next steps & whether it is still something our product team is open to accepting?

**Comment by KuSh** [comment]
> Do you have clarity on next steps & whether it is still something our product team is open to accepting? Sounded like there were some questions around it, so thought to confirm :)

It seems so yes. I've got positive feedback from @Willyfrog and @isacikgoz. PR is ready I'm waiting for the next steps

**Comment by lieut-data** [comment]
Thanks for reporting this issue and for the community discussion around it. After multiple years of inactivity, we're closing this to help keep our issue tracker focused on active work. If this is still relevant, please feel free to open a new issue with updated context and we'll be happy to take another look.
